src/redisearch_rs/slots_tracker/src/slot_set.rs

```diff
@@ -0,0 +1,1040 @@
+/*
+ * Copyright (c) 2006-Present, Redis Ltd.
+ * All rights reserved.
+ *
+ * Licensed under your choice of the Redis Source Available License 2.0
+ * (RSALv2); or (b) the Server Side Public License v1 (SSPLv1); or (c) the
+ * GNU Affero General Public License v3 (AGPLv3).
+*/
+
+//! Internal implementation of the SlotSet data structure.
+//!
+//! This module contains the private implementation of slot range tracking.
+//! It is not exposed outside of the slots_tracker crate.
+
+use crate::SlotRange;
+
+/// Enum describing the relationship between a set and a query.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub(crate) enum CoverageRelation {
+    Equals,  // Set == query (exact match)
+    Covers,  // Set ⊇ query (covers but has extra)
+    NoMatch, // Set ⊉ query (doesn't cover all)
+}
+
+// ============================================================================
+// Debug assertion helpers
+// ============================================================================
+
+/// Validates that all ranges are valid.
+#[inline]
+fn debug_assert_valid_ranges(ranges: &[SlotRange]) {
+    debug_assert!(ranges.iter().all(|r| r.start <= r.end && r.end <= 16383));
+}
+
+/// Validates that ranges are sorted and normalized (no overlaps or adjacent ranges).
+#[inline]
+fn debug_assert_normalized_ranges(ranges: &[SlotRange]) {
+    debug_assert!(ranges.windows(2).all(|w| w[0].end + 1 < w[1].start));
+}
+
+/// Validates that ranges are valid, sorted, and normalized.
+#[inline]
+fn debug_assert_valid_normalized_input(ranges: &[SlotRange]) {
+    debug_assert_valid_ranges(ranges);
+    debug_assert_normalized_ranges(ranges);
+}
```

Comment:
Why have this? It feels like rather repeating the code at the call sites will make it clearer what is being asserted